transport/httpcommon/diag.go

```diff
@@ -0,0 +1,110 @@
+	"crypto/tls"
+	"log"
+	"net/http"
+	"net/http/httptrace"
+	"net/textproto"
+)
+
+// DiagRequest returns a diagnostics hook callback that will send the passed requests using a roundtripper generated from the settings and log httptrace events in the returned bytes.
+func (settings *HTTPTransportSettings) DiagRequests(reqs []*http.Request, opts ...TransportOption) func() []byte {
+	if settings == nil {
+		return func() []byte {
+			return []byte(`error: nil httpcommon.HTTPTransportSettings`)
+		}
+	}
+	if len(reqs) == 0 {
+		return func() []byte {
+			return []byte(`error: 0 requests`)
+		}
+	}
+	return func() []byte {
+		var b bytes.Buffer
+		rt, err := settings.RoundTripper(opts...)
+		if err != nil {
+			b.WriteString("unable to get roundtripper: " + err.Error())
+			return b.Bytes()
+		}
+		logger := log.New(&b, "", log.LstdFlags|log.Lmicroseconds|log.LUTC)
+		if settings.TLS == nil {
+			logger.Print("No TLS settings")
+		} else {
+			logger.Print("TLS settings detected")
+		}
+		logger.Printf("Proxy disable=%v url=%s", settings.Proxy.Disable, settings.Proxy.URL)
+
+		ct := &httptrace.ClientTrace{
+			GetConn: func(hostPort string) {
+				logger.Printf("GetConn called for %q", hostPort)
+			},
+			GotConn: func(connInfo httptrace.GotConnInfo) {
+				logger.Printf("GotConn for %q", connInfo.Conn.RemoteAddr())
+			},
+			GotFirstResponseByte: func() {
+				logger.Print("Response started")
+			},
+			Got1xxResponse: func(code int, header textproto.MIMEHeader) error {
+				logger.Printf("Got info response status=%d, headers=%v", code, header)
+				return nil
+			},
+			DNSStart: func(info httptrace.DNSStartInfo) {
+				logger.Printf("Starting DNS lookup for %q", info.Host)
+			},
+			DNSDone: func(info httptrace.DNSDoneInfo) {
+				logger.Printf("Done DNS lookup: %+v", info)
+			},
+			ConnectStart: func(network, addr string) {
+				logger.Printf("Connection started to %s:%s", network, addr)
+			},
+			ConnectDone: func(network, addr string, err error) {
+				logger.Printf("Connection to %s:%s done, err: %v", network, addr, err)
+			},
+			TLSHandshakeStart: func() {
+				logger.Print("TLS handshake starting")
+			},
+			TLSHandshakeDone: func(state tls.ConnectionState, err error) {
+				logger.Printf("TLS handshake done. state=%+v err=%v", state, err)
+			},
+			WroteHeaders: func() {
+				logger.Printf("Wrote request headers")
+			},
+			Wait100Continue: func() {
+				logger.Printf("Waiting for continue")
+			},
+			WroteRequest: func(info httptrace.WroteRequestInfo) {
+				logger.Printf("Wrote request err=%v", info.Err)
+			},
+		}
+		for i, req := range reqs {
+			logger.Printf("Request %d to %s starting", i, req.URL.String())
+			req = req.WithContext(httptrace.WithClientTrace(req.Context(), ct))
+			if resp, err := rt.RoundTrip(req); err != nil {
+				logger.Printf("request %d error: %v", i, err)
+			} else {
+				resp.Body.Close()
+				logger.Printf("request %d successful. status=%d", i, resp.StatusCode)
+			}
+		}
+		return b.Bytes()
```

Comment:
As seen in the demo fleet-server PR the output would look like:

```
2024/05/29 00:18:41.571489 No TLS settings
2024/05/29 00:18:41.571492 Proxy disable=false url=<nil>
2024/05/29 00:18:41.571541 Request 0 to https://1df578882a2d4d0190aabbbe4ac77268.us-central1.gcp.qa.cld.elstc.co:443 starting
2024/05/29 00:18:41.571576 GetConn called for "1df578882a2d4d0190aabbbe4ac77268.us-central1.gcp.qa.cld.elstc.co:443"
2024/05/29 00:18:41.710378 Connection started to tcp:34.71.9.202:443
2024/05/29 00:18:41.834875 Connection to tcp:34.71.9.202:443 done, err: <nil>
2024/05/29 00:18:41.984139 TLS handshake starting
2024/05/29 00:18:41.984148 TLS handshake done. state={Version:772 HandshakeComplete:true DidResume:false CipherSuite:4865 NegotiatedProtocol: NegotiatedProtocolIsMutual:true ServerName:1df578882a2d4d0190aabbbe4ac77268.us-central1.gcp.qa.cld.elstc.co PeerCertificates:[0xc00021b600 0xc00021bb80 0xc00021c100] VerifiedChains:[] SignedCertificateTimestamps:[] OCSPResponse:[] TLSUnique:[] ekm:0x5557d33c1f20} err=<nil>
2024/05/29 00:18:41.984267 GotConn for "34.71.9.202:443"
2024/05/29 00:18:41.984371 Wrote request headers
2024/05/29 00:18:41.984385 Wrote request err=<nil>
2024/05/29 00:18:42.111030 Response started
2024/05/29 00:18:42.111184 request 0 successful. status=401
```
The goal of this diagnostics is to make sure that we are able to connect to the endpoint, not what that endpoint actually returns (status code is included for information, in my example i didn't set the auth header)

Reply:
Can you get any of the cert info that is pointed to by `PeerCertificates:[0xc00021b600 0xc00021bb80 0xc00021c100]` in the example?  similar to what you have cert & ca in the next example.

Reply:
Forgive the typo
```
2024/05/29 22:10:06.844374 No TLS settings
2024/05/29 22:10:06.844426 Proxy disable=false url=<nil>
2024/05/29 22:10:06.844562 Request 0 to https://3c965949205c4e3cb943d1765bb4ced0.us-central1.gcp.qa.cld.elstc.co:443 starting
2024/05/29 22:10:06.844693 GetConn called for "3c965949205c4e3cb943d1765bb4ced0.us-central1.gcp.qa.cld.elstc.co:443"
2024/05/29 22:10:06.926192 Connection started to tcp:34.71.9.202:443
2024/05/29 22:10:07.044556 Connection to tcp:34.71.9.202:443 done, err: <nil>
2024/05/29 22:10:07.222023 TLS handshake starting
2024/05/29 22:10:07.222056 TLS handshake done. state={Version:772 HandshakeComplete:true DidResume:false CipherSuite:4865 NegotiatedProtocol: NegotiatedProtocolIsMutual:true ServerName:3c965949205c4e3cb943d1765bb4ced0.us-central1.gcp.qa.cld.elstc.co PeerCertificates:[0xc0004b6b00 0xc0004b7080 0xc0004b7600] VerifiedChains:[] SignedCertificateTimestamps:[] OCSPResponse:[] TLSUnique:[] ekm:0x561fc2929f20} err=<nil>
2024/05/29 22:10:07.222167 Peer certificate count 3
2024/05/29 22:10:07.222392 - Peer Certificate 0
	Subject=CN=*.us-central1.gcp.qa.cld.elstc.co
	Issuer=CN=R3,O=Let's Encrypt,C=US
	IsCA=false
	BasicConstrintsValid=true
	NotBefore=2024-05-05 23:35:35 +0000 UTC
	NotAfter=2024-08-03 23:35:34 +0000 UTC
	Fingerprint=dr41bmmfovR/W9yZCK9EZ00RbjlwcVc2ivxaaA3FkfI=
	SAN IP=[]
	SAN DNS=[*.apm.psc.us-central1.gcp.qa.cld.elstc.co *.apm.us-central1.gcp.qa.cld.elstc.co *.app-search.psc.us-central1.gcp.qa.cld.elstc.co *.app-search.us-central1.gcp.qa.cld.elstc.co *.ent-search.psc.us-central1.gcp.qa.cld.elstc.co *.ent-search.us-central1.gcp.qa.cld.elstc.co *.ent.psc.us-central1.gcp.qa.cld.elstc.co *.ent.us-central1.gcp.qa.cld.elstc.co *.es.psc.us-central1.gcp.qa.cld.elstc.co *.es.us-central1.gcp.qa.cld.elstc.co *.fleet.psc.us-central1.gcp.qa.cld.elstc.co *.fleet.us-central1.gcp.qa.cld.elstc.co *.kb.psc.us-central1.gcp.qa.cld.elstc.co *.kb.us-central1.gcp.qa.cld.elstc.co *.profiling.psc.us-central1.gcp.qa.cld.elstc.co *.profiling.us-central1.gcp.qa.cld.elstc.co *.psc.us-central1.gcp.qa.cld.elstc.co *.symbols.psc.us-central1.gcp.qa.cld.elstc.co *.symbols.us-central1.gcp.qa.cld.elstc.co *.us-central1.gcp.qa.cld.elstc.co]
	SAN URI=[]
2024/05/29 22:10:07.223034 - Peer Certificate 1
	Subject=CN=R3,O=Let's Encrypt,C=US
	Issuer=CN=ISRG Root X1,O=Internet Security Research Group,C=US
	IsCA=true
	BasicConstrintsValid=true
	NotBefore=2020-09-04 00:00:00 +0000 UTC
	NotAfter=2025-09-15 16:00:00 +0000 UTC
	Fingerprint=jQJTbIh0grw0/1TkHSumWb+Fs0Ggogr621gT3PvPKG0=
	SAN IP=[]
	SAN DNS=[]
	SAN URI=[]
2024/05/29 22:10:07.223097 - Peer Certificate 2
	Subject=CN=ISRG Root X1,O=Internet Security Research Group,C=US
	Issuer=CN=ISRG Root X1,O=Internet Security Research Group,C=US
	IsCA=true
	BasicConstrintsValid=true
	NotBefore=2015-06-04 11:04:38 +0000 UTC
	NotAfter=2035-06-04 11:04:38 +0000 UTC
	Fingerprint=C5+lpZ7tcVwmwQIMcRtPbsQtWLABXhQzejna0wHFr8M=
	SAN IP=[]
	SAN DNS=[]
	SAN URI=[]
2024/05/29 22:10:07.223100 Verified chains count: 0
2024/05/29 22:10:07.223565 GotConn for "34.71.9.202:443"
2024/05/29 22:10:07.224495 Wrote request headers
2024/05/29 22:10:07.224507 Wrote request err=<nil>
2024/05/29 22:10:07.332675 Response started
2024/05/29 22:10:07.333657 request 0 successful. status=401
```